packages/graphql/src/classes/Subgraph.ts

```diff
@@ -0,0 +1,291 @@
+import { Executor } from "./Executor";
+
+type FederationDirectiveName =
+    | "key"
+    | "extends"
+    | "shareable"
+    | "inaccessible"
+    | "override"
+    | "external"
+    | "provides"
+    | "requires"
+    | "tag";
+
+type FullyQualifiedFederationDirectiveName = `federation__${FederationDirectiveName}`;
+
+const isFederationDirectiveName = (name: string): name is FederationDirectiveName =>
+    ["key", "extends", "shareable", "inaccessible", "override", "external", "provides", "requires", "tag"].includes(
+        name
+    );
+
+export class Subgraph {
+    private importArgument: Map<
+        FederationDirectiveName,
+        FederationDirectiveName | FullyQualifiedFederationDirectiveName | string
+    >;
+    private typeDefs: TypeSource;
+    private linkExtension: SchemaExtensionNode;
+
+    constructor(typeDefs: TypeSource) {
+        this.typeDefs = typeDefs;
+
+        this.importArgument = new Map([
+            ["key", "federation__key"],
+            ["extends", "federation__extends"],
+            ["shareable", "federation__shareable"],
+            ["inaccessible", "federation__inaccessible"],
+            ["override", "federation__override"],
+            ["external", "federation__external"],
+            ["provides", "federation__provides"],
+            ["requires", "federation__requires"],
+            ["tag", "federation__tag"],
+        ]);
+
+        const linkMeta = this.findFederationLinkMeta(typeDefs);
+        if (!linkMeta) {
+            throw new Error(`typeDefs must contain \`@link\` schema extension to be used with Apollo Federation`);
+        }
+        const { extension, directive: linkDirective } = linkMeta;
+
+        this.linkExtension = extension;
+
+        this.parseLinkImportArgument(linkDirective);
+    }
+
+    public getFullyQualifiedDirectiveName(
+        name: FederationDirectiveName
+    ): FederationDirectiveName | FullyQualifiedFederationDirectiveName | string {
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        return this.importArgument.get(name)!;
+    }
+
+    // TODO: Remove this function, parse `@shareable` directives and add them only on related generated types and fields
+    public augmentGeneratedSchemaDefinition(typeDefs: DocumentNode): DocumentNode {
+        const definitions: DefinitionNode[] = [];
+
+        const shareable = makeDirectiveNode(this.importArgument.get("shareable") as string, {}) as ConstDirectiveNode;
+
+        // This is a really filthy hack to apply @shareable
+        for (const definition of typeDefs.definitions) {
+            if (definition.kind === Kind.OBJECT_TYPE_DEFINITION) {
+                if (
+                    ["Query", "Mutation"].includes(definition.name.value) ||
+                    definition.name.value.endsWith("Connection") ||
+                    definition.name.value.endsWith("AggregateSelection") ||
+                    definition.name.value.endsWith("Edge")
```

Comment:
Moving this to a separate variable may provide a bit extra context on what are we trying to validate in this if.

I have no clue what's the relation between `Connection`, `AggregateSelection`, `Edge` and `Query | Mutation`